### PR TITLE
decouple record/train and add state readers

### PR DESCRIPTION
**Pull request by szha**
@piiswrong @zackchase 
changes include:
- decouple training and testing
- add training/record state getters to allow model codes to change behavior based on the states.

### CODE REVIEW

File: python/mxnet/autograd.py [hunk 1]
```diff
@@ -10,21 +10,21 @@
 from .symbol import _GRAD_REQ_MAP
 
 
-def set_recording(is_recording):
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: keep ing

File: python/mxnet/autograd.py [hunk 2]
```diff
@@ -45,6 +45,28 @@ def set_training(is_train):
...
+    curr = ctypes.c_bool()
+    check_call(_LIB.MXAutogradIsRecording(ctypes.byref(curr)))
+    return curr.value
+
+def is_training():
+    """Get status on training/not training.
+
+    Returns
+    -------
+    Current state of training/inference.
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: inference -> predicting

File: python/mxnet/autograd.py [hunk 3]
```diff
@@ -45,6 +45,28 @@ def set_training(is_train):
...
+    Returns
+    -------
+    Current state of recording.
+    """
+    curr = ctypes.c_bool()
+    check_call(_LIB.MXAutogradIsRecording(ctypes.byref(curr)))
+    return curr.value
+
+def is_training():
+    """Get status on training/not training.
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: predicting

File: python/mxnet/autograd.py [hunk 4]
```diff
@@ -54,26 +76,28 @@ class RecordingStateScope(object):
...
+            set_recording(self._prev_is_record)
+        if self._enter_is_train is not None and self._prev_is_train != self._enter_is_train:
             set_training(self._prev_is_train)
 
 
 def record(is_train=True):
-    """Returns a training scope context to be used in 'with' statement
-    and captures training code.
+    """Returns an autograd recording scope context to be used in 'with' statement
+    and captures codes that need gradients to be calculated.
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: code is not countable

File: python/mxnet/autograd.py [hunk 5]
```diff
@@ -112,6 +137,34 @@ def pause(is_train=False):
     return RecordingStateScope(False, is_train)
 
 
+def override_train():
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: set_training is enough for now. We can add train()/predict() after getting more user feedback

File: python/mxnet/autograd.py [hunk 6]
```diff
@@ -54,26 +76,28 @@ class RecordingStateScope(object):
...
+        self._enter_is_record = is_record
         self._enter_is_train = is_train
-        self._prev = None
+        self._prev_is_record = None
         self._prev_is_train = None
 
     def __enter__(self):
-        self._prev = set_recording(self._enter_state)
-        self._prev_is_train = set_training(self._enter_is_train)
+        if self._enter_is_record is not None:
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: revert for now

File: python/mxnet/autograd.py [hunk 7]
```diff
@@ -45,6 +45,28 @@ def set_training(is_train):
...
+    Returns
+    -------
+    Current state of training/inference.
+    """
+    curr = ctypes.c_bool()
+    check_call(_LIB.MXAutogradIsTraining(ctypes.byref(curr)))
+    return curr.value
+
 
 class RecordingStateScope(object):
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: we should probably hide this with _

File: python/mxnet/autograd.py (hunk 5 above)
Comment (szha) [inline python/mxnet/autograd.py]: I agree that we should only introduce features that users need. Just to elaborate on the reasoning: set_training() is currently action-based API, and users will need to manually revert the values if the change is intended for only a small subset, and thus will need to keep the previous states. Introducing train() and predict() scopes can help making this simpler since the "with" scope is a natural fit for such task. @zackchase what's your view as a user?

Comment (zackchase) [inline python/mxnet/autograd.py]: Thanks @szha for working hard to resolve these issues.
I really like that you've added setters / getters for training and recording and am relieved to see them decoupled. 

So one question is - 
Is predict =True equivalent to training=False? 
If so then we would only need to set training and I'd agree with Eric that we could hold off on adding predict() right now for fear of introducing redundancy. Still we have a tough situation for a modeler who wants at will to be able to run forward passes with dropout on vs off. It's hard to see how they can turn dropout on/off without messing with batchnorm.

For me, the main idea for having a predict(): scope would be *if the default were training=True*. I tend to think that tis should be the case. 

**Argument: If you just instantiate a Dropout layer and call forward on it, you'd expect the Dropout to be applied**

At the same time I'm cognizant of Eric's belief that this breaks from the previous way of doing things and might harm backwards compatibility in some sense.

Comment (szha) [inline python/mxnet/autograd.py]: Yes, it's equivalent, but the point is on how to use it. To use action-based API, user needs to keep track of the training/recording state by themselves, whereas a scope-based API does it for you. This is the difference between a sequence of get and save current state => set state => revert back the state afterwards, versus a simple scope.

Also, imagine that in the model codes you want to do different things based on whether it's training or predicting. You can't do that with action-based API without involving a book keeping for recording status.

**Comment by szha** [comment]
Doc for autograd can be found at http://mxnet-doc.s3-website-us-east-1.amazonaws.com/api/python/autograd.html

File: python/mxnet/autograd.py [hunk 8]
```diff
@@ -106,10 +135,41 @@ def pause(is_train=False):
...
+
+def train_mode():
+    """Returns a scope context to be used in 'with' statement
+    in which forward pass behavior is set to training mode,
+    without changing the recording states.
+
+    Example::
+
+        y = model(x)
+        with autograd.train():
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: train_mode

File: python/mxnet/autograd.py [hunk 9]
```diff
@@ -106,10 +135,41 @@ def pause(is_train=False):
...
+def predict_mode():
+    """Returns a scope context to be used in 'with' statement
+    in which forward pass behavior is set to inference mode,
+    without changing the recording states.
+
+    Example::
+
+        with autograd.record():
+            y = model(x)
+            with autograd.test():
```
Comment (piiswrong) [inline python/mxnet/autograd.py]: test_mode

File: tests/python/unittest/test_autograd.py [hunk 10]
```diff
@@ -251,18 +251,49 @@ def test_attach_grad():
...
+            assert (y.asnumpy() == x.asnumpy()).all()
+            y.backward(is_train=False)
+            assert (x.grad.asnumpy() == x.asnumpy()).all()
+
+    with record(train_mode=False):
+        assert is_recording()
+        assert not is_training()
         y = mx.nd.Dropout(x, p=0.5)
         assert (y.asnumpy() == x.asnumpy()).all()
         y.backward(is_train=False)
```
Comment (piiswrong) [inline tests/python/unittest/test_autograd.py]: change this to train_mode too
